README.md
# DOC Trailhead
This is the DOC Trailhead web application!

Interested in contributing? See the [contributions guidelines](CONTRIBUTIONS.md).

## Quick Start
To get started developing, run `npm run i` to install the dependencies and `npm run init-db` to
create the `trailhead.db` file. Then, you can run the server with `npm dun dev` and visit it by
navigating your browser to `localhost:8080`.

Development commands:
* `npm i` - install dependencies
* `npm t` - run tests
* `npm run dev` - run in development mode
* `npm start` - run in production mode (requires `.env` file)
* `npm run init-db` - create a clean version of the database, with a little bit of seed data
* `npm run lint` - run the linter
* `npm run format` - run the linter in `--fix` mode (might alter the code)

Database commands
* `npm run set-admin` - if using the prebuilt database, this will log the user in as an admin
* `npm run set-leader` - if using the prebuilt database, this will log the user in as trip leader

## Architecture
Trailhead is a NodeJS application. It uses SQLite as a datastore, Express to define routes, Nunjucks
to build HTML views, and HTMX to add interactivity to the webpage.

### Persistence
You can read more about SQLite in the appendix section below, but the main thing to know
about it is that its datastore is a single file, not a separate process. That means the NodeJS
application does all the database "stuff", not a separate application. This has two big implications
for the application:
* It can be run and tested without any other install besides `npm run i`. This substantially
  improves development speed and simplicity.
* Since database requests do not have to make a network hop, it basically removes the performance
  penalty for executing multiple small SQL queries (sometimes called the "[n+1 problem](https://www.sqlite.org/np1queryprob.html)"). This lets you write more intelligible SQL commands,
  mixed with a little bit of JS. In practice, I use a bunch of small queries and a few medium-sized
  ones.

### Views
This application does not really have a distinction between frontend and backend. The views are
written in an HTML templating language called [Nunjucks](https://mozilla.github.io/nunjucks/). When
the user requests a webpage, the HTML that they receive is built by the templating language at
runtime. The frontend as such is entirely stateless; whatever the user is "allowed" to do in the
application is built into the HTML that is returned from each route.

If you want to be fancy, this is a concept called Hypertext As The Engine of Application State
([HATEOAS](https://htmx.org/essays/hateoas/)). If you don't want to be fancy, think of this way: the
route responds with some HTML that contains links or buttons relevant to the user that requested it.
If the user is allowed to see a trip to hike Cardigan, the HTML will have an
`<a href=/trips/2>Cardigan Hike!</a>` element in it. If they're allowed to signup, it might have a
`<button type=submit>Signup!</button>` element. In this way the stateless hypertext (HTML) is
defining what the user is allowed to do, and Nunjucks make it easy(ish) to mix and match bits of
HTML.

## Deployment
Trailhead is deployed on a DigitalOcean droplet and served behind an NGINX proxy.

## Appendix
Here are some useful resources that might help you understand the tech stack better, if you're not
familiar with parts of it:

### Docs
* [Nunjucks Templating](https://mozilla.github.io/nunjucks/templating.html)
* [HTMX Reference](https://htmx.org/reference/)
* [better-sqlite3 API](https://github.com/WiseLibs/better-sqlite3/blob/master/docs/api.md)
* [Quirks, Caveats, and Gotchas In SQLite](https://www.sqlite.org/quirks.html)

### Blogs
* [How To Set Up a Node.js Application for Production on Ubuntu 16.04](https://www.digitalocean.com/community/tutorials/how-to-set-up-a-node-js-application-for-production-on-ubuntu-16-04) (this is basically the deployment setup)
* [Consider SQLite](https://blog.wesleyac.com/posts/consider-sqlite)
* [How Did REST Come To Mean The Opposite Of Rest?](https://htmx.org/essays/how-did-rest-come-to-mean-the-opposite-of-rest/)


HISTORY.md
# Trailhead History

# Why write this?
Because it's interesting! And someday someone that is not me will maintain this application and it
is very helpful to understand what circumstances and constraints led to the software that you're
suddenly in charge of. If you see something and think "I would have done that differently," you
might learn that it was done that way not because of incompetence, but because of time pressure,
customer requests, or a requirement that you're not aware of until you try to remove it and
something does wrong. Or incompetence.

Context builds understanding, and empathy.

## History
The first thing I want to note about Trailhead, is that the original developers did a good job.
Building a webservice from scratch is frustrating and time-consuming and when I was in the DALI lab
I didn't accomplish anything even remotely close to this. The fact that Trailhead is a
mission-critical application for the DOC is a testament to original team that worked on it, and
Ziray Hao's work in particular.

When I was handed this application, it consisted of a React frontend, an Express backend, and a
MongoDB database that the backend connected to via the Mongoose ORM. The primary problem affecting
the service was that over two years of use it had grown unbearably slow to use, with each page load
taking 5-15 seconds. It also suffered from the occasional database inconsistencies, such as trips
with no leaders, that left the application in a bad state and users stuck.

The root cause of these issues is that MongoDB is a poor persistence choice for this application,
and the backend server had architected various solutions to compensate, solutions that frequently
blocked the NodeJS event loop. It is, of course, possible to use MongoDB without blocking the event
loop, but a lot of the complexity of the application resulted from trying to reconnect relations
between data that should have been stored relationally in the first place.

Other problems include a wildly over-engineered frontend (different repo) that holds huge amounts of
information in state and crashes entirely if any of those fields are missing, and pretty chunky
security holes that make it possible for users without the correct level of permissions to access or
alter other user information. The most private thing we have on here is users' self-reported dietary
restrictions and medical conditions, which doesn't represent a massive risk (compared with actual
medical records, SSNs, etc) but is nonetheless something we should endeavor to protect.

## Migration Roadmap
1. Replace Mongoose queries with MongoDB query language, a systematic process that would identify
   the most inefficient queries while bringing the query structure for in line with the relational
   model instead of object-like model. This step is complete and it made the application snappy for
   users again.
2. Replace MongoDB and the document-based structure with SQLite and a relational table structure.
   Trailhead's data already relied on relations between the objects and this step explicitly encodes
   those relationships into the data structure to minimize the opportunities for inconsistencies.
   Because I haven't changed the frontend at all, this requires that I then map the DB table into a
   nested structure that the frontend is expecting. This is annoying, but it does actually use less
   code than we were using before, and it's well worth the benefits of data consistency and speed.
3. Remove the frontend entirely and serve HTML with an HTML-based API.
4. Now that we have a single deployable application with a lightweight UI, start implementing
   features!

# Migration Results
The website is now insanely fast. Most pages load in 25-30 ms.

The website now has a usable mobile view (though some improvements still very much necessary). This
is the result of a vastly simpler HTML and CSS, with as little nesting as possible. The CSS schema
is written mobile-first, with media queries at the bottom that are specific to desktop.

Removed around 15k lines of source code:

```
# Old Frontend
$ find src -type f | egrep '\.(js|scss|css|html)' | xargs wc -l | grep total
   17228 total

# Old Backend
$ find src -type f | egrep '\.(js|css|html|njs)' | xargs wc -l | grep total
    3561 total

# New Combined Application
$ find src templates static -type f | egrep '\.(js|css|sql|html|njs)$' | xargs wc -l |
grep total
    5272 total
```

Removed MongoDB hosting fees by removing Mongo.

Removed Heroku hosting feeds by removing Heroku. Most of the reason Heroku was costing $100 (!!) a
month was because the 2 professional dynos were having to recalculate the entire trips list every
single minute. It's actually a bit of shame that we're not on serverless anymore because servereless
is very easy to set up, but the resultant application simplicity is, I think, I worth the tradeoff
in setup complexity (for now).

## The Future
In addition to saving at least $300/mo in operating costs, I'm already finding that deploying new
features is a cinch. We'll see what happens next!

-- Alex



src/server.js
import express from 'express'
import cron from 'node-cron'
import morgan from 'morgan'
import nunjucks from 'nunjucks'
import bodyParser from 'body-parser'

import * as db from './services/sqlite.js'
import * as sessions from './services/sessions.js'
import * as mailer from './services/mailer.js'
import apiRouter from './router.js'
import restRouter from './rest-router.js'

import { requireAuth } from './services/authentication.js'

process.env.TZ = 'America/New_York'

const app = express()

// enable/disable http request logging
app.use(morgan('dev'))

// serve static files - move this to nginx eventually
app.use('/static', express.static('static'))
app.use('/htmx', express.static('node_modules/htmx.org/dist'))
app.use('/fullcalendar-scheduler', express.static('node_modules/fullcalendar-scheduler'))
app.use(bodyParser.urlencoded({ extended: true }))

nunjucks
  .configure('templates', {
    autoescape: false, // The SQLite rest methods already escape results
    express: app
  })
  .addGlobal('NODE_ENV', process.env.NODE_ENV)

app.set('views', '/templates/views')

app.use('/', apiRouter)
app.use('/rest', requireAuth, restRouter) // All REST methods require authentication
app.use(handleError)

// Open database connections
db.start('trailhead.db')
sessions.start('sessions.db')
process.on('exit', () => {
  db.stop()
  sessions.stop()
})
process.on('SIGHUP', () => process.exit(128 + 1))
process.on('SIGINT', () => process.exit(128 + 2))
process.on('SIGTERM', () => process.exit(128 + 15))

// START THE SERVER
// =============================================================================
const port = process.env.PORT || 8080
app.listen(port)
console.log(`Server running at http://localhost:${port}`)
console.error(`Starting up at ${new Date()}`)

const checkOutEmails = mailer.createRecurringEmailSender('trip check-out',
  db.getTripsPendingCheckOutEmail, mailer.sendCheckOutEmail, db.markCheckOutEmail)

const checkInEmails = mailer.createRecurringEmailSender('trip check-in',
  db.getTripsPendingCheckInEmail, mailer.sendCheckInEmail, db.markCheckInEmail)

const late90MinEmails = mailer.createRecurringEmailSender('trip late 90 minutes',
  db.getTripsPending90MinEmail, mailer.send90MinuteLateEmail, db.mark90MinEmail)

const late3HourEmails = mailer.createRecurringEmailSender('trip late 3 hours',
  db.getTripsPending3HourEmail, mailer.send3HourLateEmail, db.markTripLate)

/**
 * Schedules time-based emails.
 */
if (process.env.NODE_ENV !== 'development' && process.env.SCHEDULER_STATUS !== 'disabled') {
  // These wacky times are a stopgap to mitigate the connection limit throttling
  // I'll batch these properly (with precise queries) soon
  console.log('Starting scheduler')
  cron.schedule('0 1 * * *', db.markOldTripsAsPast)
  cron.schedule('10 * * * *', checkOutEmails)
  cron.schedule('20 * * * *', checkInEmails)
  cron.schedule('5,15,25,35,45,55 * * * *', late90MinEmails)
  cron.schedule('17,37,57 * * * *', late3HourEmails)
} else {
  console.log('Scheduler disabled')
}

function handleError (err, req, res, _next) {
  switch (err.name) {
    case 'BadRequestError':
      return res.status(400).send(err.message)
    default:
      console.error(`Unexpected error for ${req.method} ${req.url}, sending 500`)
      console.error(err.stack)
      console.error(req.body)
  }

  return res.status(500).send('Sorry, Trailhead experienced an error. Please reach to OPO.')
}
